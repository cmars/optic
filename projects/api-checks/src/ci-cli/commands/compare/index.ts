export * from './bulk-compare';
export * from './compare';
