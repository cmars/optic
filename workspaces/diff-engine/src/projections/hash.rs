use crate::{RfcEvent, SpecEvent};
use cqrs_core::{Aggregate, AggregateEvent, Event};
use serde_hashkey::to_key;
use sha2::{Digest, Sha256};
use std::{collections::BTreeMap, hash::Hash};

#[derive(Debug, Clone)]
pub struct SpecHashProjection {
  hasher: Sha256,
}

impl SpecHashProjection {
  pub fn to_string(&self) -> String {
    format!("{:x}", self.hasher.clone().finalize())
  }

  pub fn with_event(&mut self, evt: SpecEvent) {
    let key = to_key(&evt).unwrap();
    let key_str = serde_json::to_string(&key).expect("should be able to serialize");

    self.hasher.update(key_str);
  }
}

impl Default for SpecHashProjection {
  fn default() -> Self {
    SpecHashProjection {
      hasher: Sha256::new(),
    }
  }
}

impl Aggregate for SpecHashProjection {
  fn aggregate_type() -> &'static str {
    "spec_hash_projection"
  }
}

impl<I> From<I> for SpecHashProjection
where
  I: IntoIterator,
  I::Item: AggregateEvent<Self>,
{
  fn from(events: I) -> Self {
    let mut projection = SpecHashProjection::default();
    for event in events.into_iter() {
      projection.apply(event);
    }
    projection
  }
}

// Events
impl AggregateEvent<SpecHashProjection> for SpecEvent {
  fn apply_to(self, projection: &mut SpecHashProjection) {
    projection.with_event(self);
  }
}
