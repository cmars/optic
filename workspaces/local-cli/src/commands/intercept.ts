import { Command, flags } from '@oclif/command';
import {
  cleanupAndExit,
  CommandAndProxySessionManager,
  developerDebugLogger,
  fromOptic,
  loadPathsAndConfig,
  makeUiBaseUrl,
} from '@useoptic/cli-shared';
import colors from 'colors';
import { getCaptureId } from '../shared/git/git-context-capture';
import { ensureDaemonStarted } from '@useoptic/cli-server';
import { lockFilePath } from '../shared/paths';
import { Config } from '../config';
import { CaptureSaverWithDiffs } from '@useoptic/cli-shared/build/captures/avro/file-system/capture-saver-with-diffs';
import { EventEmitter } from 'events';
import { Client, SpecServiceClient } from '@useoptic/cli-client';
import getPort from 'get-port';
import url from 'url';
import { BrowserLaunchers } from '../shared/intercept/browser-launchers';
import { cli } from 'cli-ux';
import openBrowser from 'react-dev-utils/openBrowser';
import { IHttpInteraction } from '@useoptic/cli-shared/build/optic-types';

export default class Intercept extends Command {
  static description =
    'intercept API traffic for a given hostname ie api.example.com';

  static flags = {
    chrome: flags.boolean({
      description:
        'collect traffic to your deployed environment using Chrome network tab',
    }),
  };

  static args = [
    {
      name: 'environment',
      required: true,
    },
  ];

  async run() {
    const cwd = process.cwd();

    const { args, flags } = this.parse(Intercept);

    const { paths, config } = await loadPathsAndConfig(this);
    const environments = config.environments || {};

    if (!environments[args.environment]) {
      return this.warn(
        `No environment ${args.environment} found in optic.yml.Set one up like this:\n${exampleenv}`
      );
    }

    const envToTarget = environments[args.environment]!;

    const captureId = await getCaptureId(paths);
    const daemonState = await ensureDaemonStarted(
      lockFilePath,
      Config.apiBaseUrl
    );

    const apiBaseUrl = `http://localhost:${daemonState.port}/api`;
    developerDebugLogger(`api base url: ${apiBaseUrl}`);
    const cliClient = new Client(apiBaseUrl);
    const cliSession = await cliClient.findSession(cwd, null, captureId);
    const eventEmitter = new EventEmitter();
    const specServiceClient = new SpecServiceClient(
      cliSession.session.id,
      eventEmitter,
      apiBaseUrl
    );
    const persistenceManager = new CaptureSaverWithDiffs(
      {
        captureBaseDirectory: paths.capturesPath,
        captureId,
        shouldCollectDiffs: false,
      },
      config,
      specServiceClient
    );

    process.env.OPTIC_ENABLE_CAPTURE_BODY = 'yes';
    process.env.OPTIC_ENABLE_TRANSPARENT_PROXY = 'yes';

    const transparentProxyPort: number = await getPort({
      port: getPort.makeRange(3700, 3900),
    });

    const parsedTarget = url.parse(envToTarget.host);
    const serviceProtocol = parsedTarget.protocol || 'https:';

    const serviceConfig = {
      port:
        serviceProtocol === 'http:' // assume standard ports for targets without explicit ports
          ? 80
          : 443,
      host: parsedTarget.hostname!,
      protocol: serviceProtocol,
      basePath: '/',
    };

    const proxyConfig = {
      port: transparentProxyPort,
      host: 'localhost',
      protocol: serviceProtocol,
      basePath: '/',
    };

    let browsers: BrowserLaunchers | undefined = undefined;

    const onStarted = (fingerprint: string) => {
      browsers = new BrowserLaunchers(
        `${proxyConfig.protocol}//${proxyConfig.host}:${proxyConfig.port}`,
        envToTarget.webUI || envToTarget.host,
        fingerprint,
        this
      );

      if (flags.chrome) browsers.chrome();
    };

    let collected = 0;
    cli.action.start('Waiting for traffic...');
    cli.action.type = 'debug';
    const onSample = (sample: IHttpInteraction) => {
      collected++;
      cli.action.start(
        `Waiting for traffic...`,
        `${colors.gray(`${collected} requests collected`)}\n${
          sample.request.method
        } ${sample.request.host}${sample.request.path} | ${
          sample.response.statusCode
        }`
      );
    };

    const sessionManager = new CommandAndProxySessionManager(
      {
        hostnameFilter: `${serviceConfig.host}`,
        serviceConfig: serviceConfig,
        proxyConfig: proxyConfig,
      },
      onStarted,
      onSample
    );

    console.log(
      fromOptic(
        `Transparent proxy is running on ${serviceConfig.protocol}//localhost:${transparentProxyPort}\nMonitoring Traffic to ${envToTarget.host}`
      )
    );

    await sessionManager.run(persistenceManager);
    if (browsers) {
      browsers!.cleanup();
    }
    const uiBaseUrl = makeUiBaseUrl(daemonState);
    const uiUrl = `${uiBaseUrl}/apis/${cliSession.session.id}/review/${captureId}`;
    openBrowser(uiUrl);

    cleanupAndExit(0);
  }
}

const exampleenv = `
environments:
  production:
    host: https://api.github.com # the hostname of the API we should record traffic from
    webUI: https://github.com # the url that should open when a browser flag is passed\`
`;
