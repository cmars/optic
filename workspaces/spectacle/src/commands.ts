/*
TODO Replace with Rust constructors eventually
*/

// TODO fix any types in this file
export type CQRSCommand =
  | AddShapeType
  | AddRequestType
  | FieldFromShapeType
  | AddFieldType
  | ShapeProviderType
  | ProviderInShapeType
  | SetParameterShapeType
  | SetFieldShapeType
  | AddShapeParameterType
  | RemoveFieldType
  | AddPathParameterType
  | AddPathComponentType
  | AddContributionType
  | ShapedBodyDescriptorType
  | SetResponseBodyShapeType
  | SetRequestBodyShapeType
  | AddResponseByPathAndMethodType
  | PrunePathComponentsType;

export type AddShapeType = ReturnType<typeof AddShape>;
export type AddRequestType = ReturnType<typeof AddRequest>;
export type FieldFromShapeType = ReturnType<typeof FieldShapeFromShape>;
export type AddFieldType = ReturnType<typeof AddField>;
export type ShapeProviderType = ReturnType<typeof ShapeProvider>;
export type ProviderInShapeType = ReturnType<typeof ProviderInShape>;
export type SetParameterShapeType = ReturnType<typeof SetParameterShape>;
export type SetFieldShapeType = ReturnType<typeof SetFieldShape>;
export type AddShapeParameterType = ReturnType<typeof AddShapeParameter>;
export type RemoveFieldType = ReturnType<typeof RemoveField>;
export type AddPathParameterType = ReturnType<typeof AddPathParameter>;
export type AddPathComponentType = ReturnType<typeof AddPathComponent>;
export type AddContributionType = ReturnType<typeof AddContribution>;
export type ShapedBodyDescriptorType = ReturnType<typeof ShapedBodyDescriptor>;
export type SetResponseBodyShapeType = ReturnType<typeof SetResponseBodyShape>;
export type SetRequestBodyShapeType = ReturnType<typeof SetRequestBodyShape>;
export type AddResponseByPathAndMethodType = ReturnType<
  typeof AddResponseByPathAndMethod
>;
export type PrunePathComponentsType = ReturnType<typeof PrunePathComponents>;

export function AddShape(
  shapeId: string,
  baseShapeId: string,
  name: string = ''
) {
  return { AddShape: { shapeId, baseShapeId, name } };
}

export function AddRequest(
  httpMethod: string,
  pathId: string,
  requestId: string
) {
  return { AddRequest: { httpMethod, pathId, requestId } };
}

export function FieldShapeFromShape(fieldId: string, shapeId: string) {
  return { FieldShapeFromShape: { fieldId, shapeId } };
}

export function AddField(
  fieldId: string,
  shapeId: string,
  name: string,
  shapeDescriptor: FieldFromShapeType
) {
  return { AddField: { fieldId, shapeId, name, shapeDescriptor } };
}

export function ShapeProvider(shapeId: string) {
  return { ShapeProvider: { shapeId } };
}

export function ProviderInShape(
  shapeId: string,
  providerDescriptor: any,
  consumingParameterId: string
) {
  return {
    ProviderInShape: { shapeId, providerDescriptor, consumingParameterId },
  };
}

export function SetParameterShape(shapeDescriptor: any) {
  return { SetParameterShape: { shapeDescriptor } };
}
export function SetFieldShape(shapeDescriptor: FieldFromShapeType) {
  return { SetFieldShape: { shapeDescriptor } };
}

export function AddShapeParameter(
  shapeParameterId: string,
  shapeId: string,
  name: string
) {
  return { AddShapeParameter: { shapeParameterId, shapeId, name } };
}

export function RemoveField(fieldId: string) {
  return { RemoveField: { fieldId } };
}

export function AddPathParameter(
  pathId: string,
  parentPathId: string,
  name: string
) {
  return {
    AddPathParameter: {
      pathId,
      parentPathId,
      name,
    },
  };
}

export function AddPathComponent(
  pathId: string,
  parentPathId: string,
  name: string
) {
  return {
    AddPathComponent: {
      pathId,
      parentPathId,
      name,
    },
  };
}

export function AddContribution(id: string, key: string, value: string) {
  return {
    AddContribution: {
      id,
      key,
      value,
    },
  };
}

export function ShapedBodyDescriptor(
  httpContentType: string,
  shapeId: string,
  isRemoved: boolean = false
) {
  return {
    httpContentType,
    shapeId,
    isRemoved,
  };
}
export function SetResponseBodyShape(responseId: string, bodyDescriptor: any) {
  return {
    SetResponseBodyShape: {
      responseId,
      bodyDescriptor,
    },
  };
}
export function SetRequestBodyShape(requestId: string, bodyDescriptor: any) {
  return {
    SetRequestBodyShape: {
      requestId,
      bodyDescriptor,
    },
  };
}

export function AddResponseByPathAndMethod(
  responseId: string,
  pathId: string,
  httpMethod: string,
  httpStatusCode: number
) {
  return {
    AddResponseByPathAndMethod: {
      responseId,
      pathId,
      httpMethod,
      httpStatusCode,
    },
  };
}

export function PrunePathComponents() {
  return {
    PrunePathComponents: {},
  };
}
