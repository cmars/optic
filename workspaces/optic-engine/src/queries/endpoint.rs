use crate::commands::{EndpointCommand, SpecCommand};
use crate::events::HttpInteraction;
use crate::projections::endpoint::{Edge, EndpointProjection, Node, ROOT_PATH_ID};
use crate::projections::endpoint::{RequestBodyDescriptor, ResponseBodyDescriptor};
use crate::state::endpoint::{
  HttpMethod, HttpStatusCode, PathComponentId, PathComponentIdRef, RequestId, ResponseId,
};
use petgraph::graph::Graph;
use petgraph::visit::{
  depth_first_search, Control, DfsEvent, EdgeFilteredNeighborsDirected, Reversed,
};
use serde::Serialize;
use std::collections::{HashMap, HashSet};

pub struct EndpointQueries<'a> {
  pub endpoint_projection: &'a EndpointProjection,
}

impl<'a> EndpointQueries<'a> {
  pub fn new(endpoint_projection: &'a EndpointProjection) -> Self {
    EndpointQueries {
      endpoint_projection,
    }
  }

  fn extract_normalized_path(path: &str) -> &str {
    if path.eq("/") {
      path
    } else if path.ends_with("/") {
      &path[..path.len() - 1]
    } else {
      path
    }
  }

  pub fn resolve_interaction_path(
    &self,
    interaction: &HttpInteraction,
  ) -> Option<PathComponentIdRef> {
    self.resolve_path(&interaction.request.path)
  }

  pub fn resolve_path(&self, path: &str) -> Option<PathComponentIdRef> {
    let path = Self::extract_normalized_path(path);
    // eprintln!("{}", path);
    let mut path_components = path.split('/');
    // skip leading empty
    path_components.next();
    let mut last_resolved_path_id = Some(ROOT_PATH_ID);
    while let Some(s) = path_components.next() {
      // eprintln!("trying to match segment {}", s);
      let node_index = self
        .graph_get_index(last_resolved_path_id.unwrap())
        .expect("expected a node with node_id to exist");

      last_resolved_path_id = None;

      // first search for named path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();

        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("1 - neighbor {}", descriptor.name);
            if !descriptor.is_parameter {
              if descriptor.name == s {
                last_resolved_path_id = Some(id);
                break;
              }
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      // try path parameters since we didn't find any matches in regular path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();
        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("2 - neighbor {}", descriptor.name);

            if descriptor.is_parameter {
              last_resolved_path_id = Some(id);
              break;
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      return None;
    }
    last_resolved_path_id
  }

  pub fn resolve_unused_paths(&self) -> impl Iterator<Item = PathComponentId> + '_ {
    let root_path_node_index = self
      .graph_get_index(ROOT_PATH_ID)
      .expect("a root path component node should exist");

    let reversed_graph = Reversed(&self.endpoint_projection.graph);

    let mut path_ids_with_endpoints: HashSet<PathComponentId> = HashSet::new();
    let mut unused_path_ids: HashSet<PathComponentId> = HashSet::new();
    let mut unused_path_ids_sorted: Vec<PathComponentId> = Vec::new();
    let mut path_ids_by_index = HashMap::new();
    let mut children_path_ids_by_path_id = HashMap::new();

    depth_first_search(&reversed_graph, Some(*root_path_node_index), |event| {
      match event {
        // Discovered a new node
        DfsEvent::Discover(discovered_node_index, time) => {
          let node = self
            .endpoint_projection
            .graph
            .node_weight(discovered_node_index);
          if let Some(Node::PathComponent(path_id, _)) = node {
            path_ids_by_index.insert(discovered_node_index, path_id.clone());
            children_path_ids_by_path_id.insert(path_id.clone(), vec![]);
            Control::Continue
          } else {
            Control::Prune // Stop going down this branch when not a path component
          }
        }
        // discovered an outgoing edge from an already discovered node
        DfsEvent::TreeEdge(parent_node_index, child_node_index) => {
          let child_node = self
            .endpoint_projection
            .graph
            .node_weight(child_node_index)
            .expect("target node should exists");

          let path_id = path_ids_by_index
            .get(&parent_node_index)
            .expect("parent path node should already have been discovered");

          match child_node {
            Node::PathComponent(child_path_id, _) => {
              let children_path_ids = children_path_ids_by_path_id
                .get_mut(path_id)
                .expect("parent should have created collection of child paths");
              children_path_ids.push(child_path_id);
              Control::Continue // we'll want to discover at the target edge as well
            }
            Node::HttpMethod(method) => {
              let request_ids = self
                .resolve_requests(path_id, method)
                .expect("path should exist")
                .map(|_| ());
              let response_ids = self
                .resolve_responses(path_id, method)
                .expect("path should exist")
                .map(|_| ());

              if let Some(_) = request_ids.chain(response_ids).next() {
                // found a request or response, path is being used
                path_ids_with_endpoints.insert(path_id.clone());
              }

              Control::Prune // no more paths down this branch
            }
            _ => Control::Prune, // no more paths down this branch
          }
        }
        DfsEvent::Finish(finished_node_index, time) => {
          let is_root_node = *root_path_node_index == finished_node_index;
          let path_id = path_ids_by_index
            .get(&finished_node_index)
            .expect("finished path node should already have been discovered");

          let has_used_child = children_path_ids_by_path_id
            .get(path_id)
            .expect("collection of child paths should have been created during discovery")
            .iter()
            .any(|child_path_id| !unused_path_ids.contains(*child_path_id));

          if !path_ids_with_endpoints.contains(path_id) && !has_used_child && !is_root_node {
            unused_path_ids.insert(path_id.clone());
            unused_path_ids_sorted.push(path_id.clone());
          }

          if is_root_node {
            Control::Break(())
          } else {
            Control::Continue
          }
        }
        _ => Control::Continue,
      }
    });

    unused_path_ids_sorted.into_iter()
  }

  pub fn resolve_operations_by_request_method(
    &self,
    method: &'a String,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    self
      .resolve_requests(path_id, method)
      .expect("expected a operations to exist")
  }

  pub fn resolve_operations(
    &self,
    interaction: &'a HttpInteraction,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    self.resolve_operations_by_request_method(&interaction.request.method, path_id)
  }

  pub fn resolve_requests(
    &self,
    path_id: PathComponentIdRef,
    method: &'a String,
  ) -> Option<impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)>> {
    let path_node_index = self.graph_get_index(path_id)?;
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_method = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        match node {
          Node::HttpMethod(http_method) => method == http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);

        let operations = children.filter_map(move |i| {
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::Request(request_id, request_descriptor) => Some((request_id, request_descriptor)),
            _ => None,
          }
        });
        operations
      });
    Some(matching_method)
  }

  pub fn resolve_request_by_method_and_content_type(
    &self,
    path_id: PathComponentIdRef,
    method: &'a String,
    content_type: Option<&'a String>,
  ) -> Option<(&RequestId, &RequestBodyDescriptor)> {
    self.resolve_requests(path_id, method).and_then(|mut it| {
      it.find(|(id, body)| match content_type {
        Some(content_type) => match body.body {
          Some(ref body) => body.http_content_type.eq(content_type),
          None => false,
        },
        None => body.body.is_none(),
      })
    })
  }

  pub fn resolve_responses(
    &self,
    path_id: &'a PathComponentId,
    method: &'a String,
  ) -> Option<impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)>> {
    let response_nodes = self
      .endpoint_projection
      .get_endpoint_response_nodes(path_id, method)?;

    Some(response_nodes.map(|node| match node {
      Node::Response(response_id, body_descriptor) => (response_id, body_descriptor),
      _ => unreachable!("get response nodes should only return response nodes"),
    }))
  }

  pub fn resolve_response_by_method_status_code_and_content_type(
    &self,
    path_id: PathComponentIdRef,
    method: &'a str,
    status_code: u16,
    content_type: Option<&'a String>,
  ) -> Option<(&ResponseId, &ResponseBodyDescriptor)> {
    self
      .resolve_responses_by_method_and_status_code(method, status_code, path_id)
      .find(|(id, body)| match content_type {
        Some(content_type) => match body.body {
          Some(ref body) => body.http_content_type.eq(content_type),
          None => false,
        },
        None => body.body.is_none(),
      })
  }

  pub fn resolve_responses_by_method_and_status_code(
    &self,
    method: &'a str,
    status_code: u16,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)> {
    let path_node_index = self
      .graph_get_index(path_id)
      .expect("expected a node with node_id to exist");
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_status_code = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        // eprintln!("method node {:?}", node);
        match node {
          Node::HttpMethod(http_method) => method == *http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        // eprintln!("method node id {:?}", i);
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let status_code_nodes = children.filter_map(move |i| {
          // eprintln!("method child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::HttpStatusCode(http_status_code) => {
              // eprintln!("status code {:?}", http_status_code);
              if status_code == *http_status_code {
                Some(i)
              } else {
                None
              }
            }
            _ => None,
          }
        });
        status_code_nodes
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let response_nodes = children.filter_map(move |i| {
          // eprintln!("status_code child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          // eprintln!("status_code child node {:?}", node);
          match node {
            Node::Response(response_id, response_descriptor) => {
              Some((response_id, response_descriptor))
            }
            _ => None,
          }
        });
        response_nodes
      });
    matching_status_code
  }

  pub fn delete_endpoint_commands(
    &self,
    path_id: &'a PathComponentId,
    method: &'a HttpMethod,
  ) -> Option<DeleteEndpointCommands> {
    let request_ids = self
      .resolve_requests(path_id, method)?
      .map(|(request_id, _)| request_id);
    let response_ids = self
      .resolve_responses(path_id, method)?
      .map(|(response_id, _)| response_id);

    let request_commands = request_ids.cloned().map(EndpointCommand::remove_request);
    let response_commands = response_ids.cloned().map(EndpointCommand::remove_response);

    Some(DeleteEndpointCommands {
      path_id: path_id.clone(),
      method: method.clone(),
      commands: request_commands
        .chain(response_commands)
        .map(SpecCommand::from)
        .collect(),
    })
  }

  pub fn delete_path_commands(
    &self,
    path_id: &'a PathComponentId,
  ) -> Option<impl Iterator<Item = EndpointCommand>> {
    let path_descriptor = self
      .endpoint_projection
      .get_path_component_descriptor(path_id)?;
    let command = if path_descriptor.is_parameter {
      EndpointCommand::remove_path_parameter(path_id.clone())
    } else {
      EndpointCommand::remove_path_component(path_id.clone())
    };

    Some(std::iter::once(command))
  }

  fn graph_get_index(&self, node_id: &str) -> Option<&petgraph::graph::NodeIndex> {
    self.endpoint_projection.node_id_to_index.get(node_id)
  }

  fn graph_get_node(&self, node_id: &str) -> &Node {
    self
      .endpoint_projection
      .graph
      .node_weight(
        *self
          .graph_get_index(node_id)
          .expect("expected a node with node_id to exist"),
      )
      .expect("expected node with node_id to exist")
  }

  fn graph_get_children(
    &self,
    node_index: &petgraph::graph::NodeIndex,
  ) -> petgraph::graph::Neighbors<Edge> {
    let neighbors = self
      .endpoint_projection
      .graph
      .neighbors_directed(*node_index, petgraph::Direction::Incoming);
    return neighbors;
  }
}

#[derive(Debug, Serialize)]
pub struct DeleteEndpointCommands {
  path_id: PathComponentId,
  method: HttpMethod,
  commands: Vec<SpecCommand>,
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::events::SpecEvent;
  use crate::projections::SpecProjection;
  use crate::Aggregate;
  use insta::assert_debug_snapshot;
  use petgraph::dot::Dot;
  use serde_json::json;

  fn interaction_with_path(path: String) -> HttpInteraction {
    serde_json::from_value(json!(
      {
        "uuid": "id",
        "request": {
          "host": "example.com",
          "method": "GET",
          "path": path,
          "query": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "headers": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "body": {
            "contentType": null,
            "value": {
              "shapeHashV1Base64": null,
              "asJsonString": null,
              "asText": null
            }
          }
        },
        "response": {
          "statusCode": 200,
          "headers": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "body": {
            "contentType": null,
            "value": {
              "shapeHashV1Base64": null,
              "asJsonString": null,
              "asText": null
            }
          }
        },
        "tags": []
      }
    ))
    .unwrap()
  }
  #[test]
  pub fn can_ignore_trailing_slash() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/a/b/c/"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/a/b/c")
  }

  #[test]
  pub fn can_handle_no_trailing_slash() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/a/b/c"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/a/b/c")
  }
  #[test]
  pub fn can_handle_root_path() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/")
  }

  #[test]
  pub fn can_find_unused_paths() {
    let events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": { "pathId": "path_1", "parentPathId": "root", "name": "posts" }},
      {"PathComponentAdded": { "pathId": "path_2", "parentPathId": "path_1", "name": "favourites" }},
      {"PathComponentAdded": { "pathId": "path_3", "parentPathId": "root", "name": "authors" }},
      {"PathComponentAdded": { "pathId": "path_4", "parentPathId": "path_3", "name": "dutch" }},
      {"PathParameterAdded":{"pathId":"path_parameter_1","parentPathId":"path_1","name":"postId"}},
      {"ShapeAdded":{"shapeId":"shape_1","baseShapeId":"$string","parameters":{"DynamicParameterList":{"shapeParameterIds":[]}},"name":"" }},
      {"PathParameterShapeSet":{"pathId":"path_parameter_1","shapeDescriptor":{"shapeId":"shape_Ba53AWXhVW","isRemoved":false} }},

      {"RequestAdded": { "requestId": "request_1", "pathId": "path_2", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_1", "pathId": "path_2", "httpMethod": "GET", "httpStatusCode": 200 }},

      {"RequestAdded": { "requestId": "request_2", "pathId": "path_1", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_2", "pathId": "path_1", "httpMethod": "GET", "httpStatusCode": 200 }},

      {"RequestRemoved": { "requestId": "request_2" }},
      {"ResponseRemoved": { "responseId": "response_2" }},
    ]))
    .expect("should be able to deserialize test events");

    let spec_projection = SpecProjection::from(events);
    // dbg!(Dot::with_config(&spec_projection.endpoint().graph, &[]));

    let endpoint_queries = EndpointQueries::new(spec_projection.endpoint());

    let unused_path_ids = endpoint_queries.resolve_unused_paths().collect::<Vec<_>>();

    assert_debug_snapshot!("can_find_unused_paths__unused_paths", unused_path_ids);
  }

  #[test]
  pub fn can_generate_delete_endpoint_commands() {
    let events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": { "pathId": "path_1", "parentPathId": "root", "name": "posts" }},
      {"PathComponentAdded": { "pathId": "path_2", "parentPathId": "path_1", "name": "favourites" }},
      {"PathComponentAdded": { "pathId": "path_3", "parentPathId": "root", "name": "authors" }},

      {"RequestAdded": { "requestId": "request_1", "pathId": "path_2", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_1", "pathId": "path_2", "httpMethod": "GET", "httpStatusCode": 200 }},

      {"RequestAdded": { "requestId": "request_2", "pathId": "path_2", "httpMethod": "POST"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_2", "pathId": "path_2", "httpMethod": "POST", "httpStatusCode": 201 }},

      {"RequestAdded": { "requestId": "request_3", "pathId": "path_3", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_3", "pathId": "path_3", "httpMethod": "GET", "httpStatusCode": 200 }},
    ]))
    .expect("should be able to deserialize test events");

    let spec_projection = SpecProjection::from(events);
    // dbg!(Dot::with_config(&spec_projection.endpoint().graph, &[]));

    let endpoint_queries = EndpointQueries::new(spec_projection.endpoint());

    let subject_path = String::from("path_2");
    let subject_method = String::from("GET");
    let deleted_endpoint_commmands = endpoint_queries
      .delete_endpoint_commands(&subject_path, &subject_method)
      .expect("delete commands are generated for existing path and method");

    let updated_spec =
      assert_valid_commands(spec_projection.clone(), deleted_endpoint_commmands.commands);
    let updated_queries = EndpointQueries::new(&updated_spec.endpoint());
    let remaining_requests = updated_queries
      .resolve_requests(&subject_path, &subject_method)
      .unwrap()
      .collect::<Vec<_>>();
    let remaining_responses = updated_queries
      .resolve_responses(&subject_path, &subject_method)
      .unwrap()
      .collect::<Vec<_>>();

    // dbg!(Dot::with_config(&updated_spec.endpoint().graph, &[]));

    assert_eq!(remaining_requests.len(), 0);
    assert_eq!(remaining_responses.len(), 0);
  }

  #[test]
  pub fn can_generate_delete_path_commands() {
    let events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": { "pathId": "path_1", "parentPathId": "root", "name": "posts" }},
      {"PathComponentAdded": { "pathId": "path_2", "parentPathId": "path_1", "name": "favourites" }},
      {"PathComponentAdded": { "pathId": "path_3", "parentPathId": "root", "name": "authors" }},

      {"PathParameterAdded":{"pathId":"path_parameter_1","parentPathId":"path_1","name":"postId"}},
      {"ShapeAdded":{"shapeId":"shape_1","baseShapeId":"$string","parameters":{"DynamicParameterList":{"shapeParameterIds":[]}},"name":"" }},
      {"PathParameterShapeSet":{"pathId":"path_parameter_1","shapeDescriptor":{"shapeId":"shape_Ba53AWXhVW","isRemoved":false} }},
    ]))
    .expect("should be able to deserialize test events");

    let spec_projection = SpecProjection::from(events);
    // dbg!(Dot::with_config(&spec_projection.endpoint().graph, &[]));

    let endpoint_queries = EndpointQueries::new(spec_projection.endpoint());

    let subjects = [("path_3", "/authors"), ("path_parameter_1", "/posts/1")];

    let delete_path_commands = subjects
      .iter()
      .flat_map(|(subject_path_id, _)| {
        endpoint_queries
          .delete_path_commands(&String::from(*subject_path_id))
          .expect("delete commands are generated for existing path")
      })
      .map(|endpoint_command| SpecCommand::from(endpoint_command))
      .collect::<Vec<_>>();

    let updated_spec = assert_valid_commands(spec_projection.clone(), delete_path_commands);
    let updated_queries = EndpointQueries::new(&updated_spec.endpoint());
    let remaining_paths = subjects
      .iter()
      .filter_map(|(_, subject_path)| updated_queries.resolve_path(*subject_path))
      .collect::<Vec<_>>();

    // dbg!(Dot::with_config(&updated_spec.endpoint().graph, &[]));

    // TODO: enable assertion once projection is updated
    assert_eq!(remaining_paths.len(), 0);
  }

  fn assert_valid_commands(
    mut spec_projection: SpecProjection,
    commands: impl IntoIterator<Item = SpecCommand>,
  ) -> SpecProjection {
    // let mut spec_projection = SpecProjection::default();
    for command in commands {
      let events = spec_projection
        .execute(command)
        .expect("generated commands must be valid");

      for event in events {
        spec_projection.apply(event)
      }
    }

    spec_projection
  }
}
