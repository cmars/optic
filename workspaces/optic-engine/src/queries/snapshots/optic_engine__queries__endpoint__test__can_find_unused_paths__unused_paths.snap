---
source: workspaces/optic-engine/src/queries/endpoint.rs
expression: unused_path_ids
---
[
    "path_4",
    "path_3",
    "path_parameter_1",
]
